teps/0095-common-repository-configuration.md

```diff
@@ -0,0 +1,276 @@
+repo configuration only once, rather than per-project. Over time other projects
+and tools could also leverage this information.
+
+## Motivation
+
+Source repositories hold a unique position in CI/CD, not only as
+versioned storage for code but also as a source of events, target of
+notifications like test status updates, source of tasks and pipelines,
+"single pane of glass" for developers, and so on. It therefore makes
+sense that several Tekton components are designed to work directly with
+source repos, either reading or writing to them in some way.
+
+The core motivation for this TEP is to model source repos explicitly in
+Tekton's domain so that components that work with them have a shared
+set of configuration describing them.
+
+### Goals
+
+The precise scope of this TEP depends on results of experimenting with
+this design prior to defining a proposal.
+
+- Decide if there is a common subset of repository data that would be
+  useful across multiple projects.
+- Design a format to store and record that common data.
+- Publish developer-facing docs for projects to use.
+- Publish user-facing docs for operators/cluster admins/app teams to use.
+
+### Non-Goals
+
+- Mutable `status`. This TEP is not aiming to offer a way for concurrent
+  processess to modify shared state or communicate about state changes.
+  Components can read and respond to the shared Repository configuration
+  data but shouldn't expect access to write back to it.
+- A standard for inferring or storing provenance information related to
+  source code or repositories. This kind of data modelling is kept to
+  the domain of individual components rather than mandated by the common
+  repository configuration.
+- Defining interfaces between components for repository data, for
+  example in Pipeline params. Instead this TEP can lean on the work from
+  the dictionary params TEP, which sets the way for those interfaces to
+  be defined.
+
+### Use Cases
+
+These use-cases were primarily taken from [the Workflows WG meeting
+notes](https://docs.google.com/document/d/1di4ikeVb8Mksgbq4CzW4m4xUQPZ2dQMLvK1VIJw7OQg/edit#heading=h.i22qgrdiutdu)
+on Nov 16, 2021.
+
+- A component could read the type, provider and API endpoint info from
+  the common repo config to determine how to convert `CloudEvents` from
+  Pipelines into status updates on that repo.
+- A component could record history related to a Repository object (e.g.
+  the set of PipelineRuns that have been started as a result of events from
+  that Repo) and then provide services based on that history (e.g. "clean up
+  all the PipelineRuns started for this Pull Request on that Repo" or
+  "return a summary of all PipelineRuns started for this repo in the
+  past 7 days").
+- A UI component could look up the website URL to link to in order
+  to view detailed pull request information in that repo.
+- A component could observe Repository objects and automate setting up
+  `EventListeners` and `Triggers` for it.
```

Comment:
If this is done via reconciliation, it would probably make sense to add a `Status` to the repo, which declares that the resources have been provisioned for it. It could even be a webhook provisioning service.
The user, but looking at the repo resource would know whether the webhook has been provisioned and what its URL is. 

This would go against the "non-goal" of `mutable` status though.
Perhaps a webhook provisioner could have its own repo CRD, automatically created once a repo configuration is discovered, and store status there?